include/ipc/ipc_service.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_IPC_SERVICE_IPC_SERVICE_H_
+#define ZEPHYR_INCLUDE_IPC_SERVICE_IPC_SERVICE_H_
+
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief IPC Service API
+ * @defgroup ipc_service_api IPC service APIs
+ * @{
+ */
+
+/** @brief Event callback structure
+ *  It is registered during endpoint registration.
+ *  This structure is packed into the endpoint configuration.
+ */
+struct ipc_service_cb {
+	/** @brief Bind was successful
+	 *  @param priv Private user data.
+	 */
+	int (*bound)(void *priv);
+
+	/** @brief New packet arrived.
+	 *  @param data Pointer to data buffer.
+	 *  @param len Length of @a data.
+	 *  @param priv Private user data.
+	 */
+	int (*received)(void *data, size_t len, void *priv);
+
+	/** @brief An error occurred
+	 *  @param message Error message.
+	 *  @param priv Private user data.
+	 */
+	int (*error)(const char *message, void *priv);
+};
+
+/** @brief Endpoint instance
+ *  Content is not important for user of the API.
+ *  It is implemented in a specific backend.
+ */
+struct ipc_ept;
+
+/** @brief Endpoint configuration structure */
+struct ipc_ept_cfg {
+	const char *name;           /**< @brief Name of the endpoint */
+	struct ipc_service_cb cb;   /**< @brief Event callback structure */
+	int prio;                   /**< @brief Endpoint priority */
+							    /* If the backend supports priorities*/
+	void *priv;                 /**< @brief Private user data */
+};
```

Comment:
In this proposal, endpoints are identified only by service name `name`. It is fine and precise enough in two core systems. However, to make this solution generic we would need to extend this identification.

From Linux rpmsg documentation:

> OMAP4, for example, has dual Cortex-A9, dual Cortex-M3 and a C64x+ DSP.

To support such more complex architectures, alongside service name, API user should select which processor is the destination of given IPC endpoint. The simplest solution would be to add one more field in this structure to identify target processor (I'm not sure if there is processor identification abstraction in Zephyr yet). Another one would be to replace `const char *name;` with `struct ipc_ipt_id id;`. This new structure could contain only one field (`name`) for now and could be extended later with target processor identification, when any backend would support multiple target processors.

Reply:
That' s right, this API proposal is sufficient for two cores, but not for more.
It appears that it will be necessary to add a field to the structure specifying the target processor.  
This information along with the name and priority of the endpoint can be placed in an additional structure. for example:
```
struct ipc_ept_id {
    const char *name;
    int target_core_id;
    int prio;
};

struct ipc_ept_cfg {
    struct ipc_ept_id ept_id;
    struct ipc_service_cb cb;  
    void *priv;
};           
```
Is there a mechanism to identify processors in Zephyr?